File: cs/src/Connections/LiveShareRelayTunnelClient.cs
// <copyright file="LiveShareRelayTunnelClient.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license.
// </copyright>

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Azure.Relay;
using Microsoft.VisualStudio.Ssh;
using Microsoft.VsSaaS.TunnelService.Contracts;

namespace Microsoft.VsSaaS.TunnelService
{
    /// <summary>
    /// Tunnel client implementation that connects via a Live Share session's Azure Relay.
    /// </summary>
    public class LiveShareRelayTunnelClient : TunnelClientBase
    {
        /// <summary>
        /// Creates a new instance of a client that connects to a tunnel via a Live Share
        /// session's Azure Relay.
        /// </summary>
        public LiveShareRelayTunnelClient(
            ITunnelManagementClient managementClient,
            TraceSource trace) : base(trace)
        {
            ManagementClient = Requires.NotNull(managementClient, nameof(managementClient));
        }

        private ITunnelManagementClient ManagementClient { get; }

        /// <inheritdoc />
        public override IReadOnlyCollection<TunnelConnectionMode> ConnectionModes
            => new[] { TunnelConnectionMode.LiveShareRelay };

        /// <inheritdoc />
        protected override async Task ConnectAsync(Tunnel tunnel, IEnumerable<TunnelEndpoint> endpoints, CancellationToken cancellation)
        {
            var liveShareEndpoint = endpoints
                .OfType<LiveShareRelayTunnelEndpoint>()
                .SingleOrDefault() ??
                throw new InvalidOperationException(
                    "The host is not currently accepting Live Share relay connections.");

            var relayUri = liveShareEndpoint.RelayUri;
            if (string.IsNullOrEmpty(relayUri))
            {
                throw new InvalidOperationException(
                    "The Live Share relay endpoint URI is missing.");
            }

            var tokenProvider = new LiveShareRelayTokenProvider(
                ManagementClient,
                tunnel,
                liveShareEndpoint.HostId!,
                TunnelAccessScopes.Connect,
                liveShareEndpoint.RelayClientSasToken);
            var relayClient = new HybridConnectionClient(new Uri(relayUri), tokenProvider);

            HybridConnectionStream stream;
            try
            {
                stream = await relayClient.CreateConnectionAsync();
            }
            catch (RelayException rex)
            {
                throw new TunnelConnectionException("Failed to connect to tunnel relay.", rex);
            }

            try
            {
                await StartSshSessionAsync(stream, cancellation);
            }
            catch (SshConnectionException scex)
            {
                throw new TunnelConnectionException("Failed to secure tunnel connection.", scex);
            }
        }
    }
}

File: ts/src/connections/liveShareRelayTunnelClient.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { SshClientSession, Trace } from '@vs/vs-ssh';
import { SshHelpers } from './sshHelpers';
import { TunnelClientBase } from './tunnelClientBase';
import {
    Tunnel,
    TunnelConnectionMode,
    TunnelEndpoint,
    LiveShareRelayTunnelEndpoint,
} from '@vs/tunnels-contracts';

/**
 * Tunnel client implementation that connects via a Live Share session's Azure Relay.
 */
export class LiveShareRelayTunnelClient extends TunnelClientBase {
    public sshSession?: SshClientSession;
    public connectionModes = [TunnelConnectionMode.LiveShareRelay];
    public trace: Trace = (level, eventId, msg, err) => {};

    constructor() {
        super();
    }

    public async connectClient(tunnel: Tunnel, endpoints: TunnelEndpoint[]): Promise<void> {
        let liveShareEndpoints = endpoints.map(
            (endpoint) => endpoint as LiveShareRelayTunnelEndpoint,
        );
        let liveShareEndpoint;
        if (liveShareEndpoints && liveShareEndpoints.length === 1) {
            liveShareEndpoint = liveShareEndpoints[0];
        } else {
            throw new Error('The host is not currently accepting Live Share relay connections.');
        }

        let relayUri = liveShareEndpoint.relayUri;
        if (!relayUri) {
            throw new Error('The Live Share relay endpoint URI is missing.');
        }

        const url = SshHelpers.getRelayUri(
            relayUri,
            liveShareEndpoint.relayClientSasToken,
            'connect',
        );
        try {
            const stream = await SshHelpers.openConnection(url);
            await this.startSshSession(stream);
        } catch (ex) {
            console.log(ex);
        }
    }
}
